openmc/data/resonance_covariance.py

```diff
@@ -0,0 +1,812 @@
+                               'Covariance values for the unresolved region not imported.'
+                    warnings.warn(warn_str)
+                                 
+
+        return cls(ranges)
+
+
+class ResonanceCovarianceRange:
+    """Resonace covariance range. Base class for different formalisms.
+
+    Parameters
+    ----------
+    energy_min : float
+        Minimum energy of the resolved resonance range in eV
+    energy_max : float
+        Maximum energy of the resolved resonance range in eV
+
+    Attributes
+    ----------
+    energy_min : float
+        Minimum energy of the resolved resonance range in eV
+    energy_max : float
+        Maximum energy of the resolved resonance range in eV
+    parameters : pandas.DataFrame
+        Resonance parameters
+    covariance : numpy.array
+        The covariance matrix contained within the ENDF evaluation
+    lcomp : int
+        Flag indicating the format of the covariance matrix within the ENDF file
+    mpar : int
+        Number of parameters in covariance matrix for each individual resonance
+    formalism : str
+        String descriptor of formalism
+    """
+    def __init__(self, energy_min, energy_max):
+        self.energy_min = energy_min
+        self.energy_max = energy_max
+    
+    def res_subset(self, parameter_str, bounds):
+        """Produce a subset of resonance parameters and the corresponding
+        covariance matrix to an IncidentNeutron object.
+        
+        Parameters
+        ----------
+        parameter_str : str
+            parameter to be discriminated
+            (i.e. 'energy', 'captureWidth', 'fissionWidthA'...)
+        bounds : np.array 
+            [low numerical bound, high numerical bound]
+    
+        Returns
+        -------
+        parameters_subset : pandas.Dataframe
+            Subset of parameters (maintains indexing of original)
+        cov_subset : np.array
+            Subset of covariance matrix (upper triangular)
+        
+        """
+        parameters = self.parameters
+        cov = self.covariance
+        mpar = self.mpar
+        mask1 = parameters[parameter_str] >= bounds[0]
+        mask2 = parameters[parameter_str] <= bounds[1]
+        mask = mask1 & mask2
+        parameters_subset = parameters[mask] 
+        indices = parameters_subset.index.values
+        sub_cov_dim = len(indices)*mpar
+        cov_subset_vals = []
+        for index1 in indices:
+            for i in range(mpar):
+                for index2 in indices:
+                    for j in range(mpar):
+                        if index2*mpar+j >= index1*mpar+i:
+                            cov_subset_vals.append(cov[index1*mpar+i, index2*mpar+j])
+    
+        cov_subset = np.zeros([sub_cov_dim, sub_cov_dim])
+        tri_indices = np.triu_indices(sub_cov_dim)
+        cov_subset[tri_indices] = cov_subset_vals
+    
+        self.parameters_subset = parameters_subset
+        self.cov_subset = cov_subset
```

Comment:
As with the sampling method, this method should return an object rather than setting attributes of the current instance. I think what would make the most sense is to return a new instance of `ResonanceCovarianceRange` with the specified subset of parameters/covariance matrix. This would allow you to get rid of the `use_subset` argument on the sampling method since calling `sample_resonance_parameters` on the new instance would use the subset by default.